deltacat/compute/compactor_v2/steps/merge.py

```diff
@@ -257,14 +257,170 @@ def _copy_all_manifest_files_from_old_hash_buckets(
     return materialize_result_list
 
 
+def _does_hash_bucket_idx_have_compacted_table(input: MergeInput, hb_idx: int) -> bool:
+    return (
+        input.round_completion_info
+        and input.round_completion_info.hb_index_to_entry_range
+        and input.round_completion_info.hb_index_to_entry_range.get(str(hb_idx))
+        is not None
+    )
+
+
+def _flatten_dfe_list(
+    df_envelopes_list: List[List[DeltaFileEnvelope]],
+) -> List[DeltaFileEnvelope]:
+    if not df_envelopes_list:
+        return []
+    return [d for dfe_list in df_envelopes_list for d in dfe_list]
+
+
+def _sort_df_envelopes(
+    df_envelopes: List[DeltaFileEnvelope],
+    key: Callable = lambda df: (df.stream_position, df.file_index),
+) -> List[DeltaFileEnvelope]:
+    if not df_envelopes:
+        return []
+    return sorted(
+        df_envelopes,
+        key=key,
+        reverse=False,  # ascending
+    )
+
+
+def group_by_upsert_delete_sequence(
+    df_envelopes: List[DeltaFileEnvelope],
+) -> Iterator[Tuple[List, List]]:
+    """
+    TODO
+    """
+    iter_df_envelopes = iter(df_envelopes)
+    upserts, deletes = [], []
+    for is_upsert, group_envelopes in itertools.groupby(
+        iter_df_envelopes, lambda x: x.delta_type == DeltaType.UPSERT
+    ):
+        if is_upsert:
+            upserts.extend(group_envelopes)
+        else:
+            deletes.extend(group_envelopes)
+        if deletes or not (deletes or upserts):
```

Comment:
`not (deletes or upserts)` is true when both are empty. I don't understand this.